Skip move logic if the parent is staying the same

CyberReiter opened the pull request:
### Prerequisites

- [x] I have added steps to test this contribution in the description below

### Description
Added checks to ContentTypes, Content and DataTypes Move functions to skip all the logic if the parent would stay the same.

### How to test
Try moving around ContentTypes, Content and DataTypes and make sure this still works as expected

mikecp commented:
Hello @CyberReiter ,

Thanks for this update, it's always better to prevent unnecessary actions 😉

I was wondering if you could also implement this on the Dictionary items move action, which is located in  `Umbraco.Cms.Web.BackOffice.Controllers.DictionaryController`, in the `PostMove `method.
It would be great if we could have the same behavior there as well 😁

Cheers!

CyberReiter commented:
Hey @mikecp 

Good catch, I totally forgot to also check Controllers for such logic!
Added some checks and also updated the translations to reflect this changes

mikecp commented:
Hi @CyberReiter ,

Thanks for the update, it's good that we cover dictionary items as well 👍

I have one last remark, and I'm very sorry I did not notice it the first time. While testing today, I noticed that when we move media items and dictionary items under the same parent, we get an error message, while we get a success message in the other cases.

I think it would be more consistent to have the same behavior in all cases, and I guess it is then better to align with what already existed, which would then be to display errors in the other situations.
From what I saw, it can be adapted in the content type and data type logic by replacing the return line

`return OperationResult.Attempt.Succeed(MoveOperationStatusType.Success, eventMessages);`

with something like

`return OperationResult.Attempt.Fail(MoveOperationStatusType.FailedNotAllowedByPath, eventMessages);`

We then get an error like this, which is similar to what we get on Dictionary:

![image](https://user-images.githubusercontent.com/1769475/188760344-80684518-3be9-4a24-b1a4-7f44cef2066d.png)

For the moving of content itself, the logic is quite different and it seems not possible to return an error without creating a full scope, so here I would leave it as you implemented it, and return a success. I'm pinging @nul800sebastiaan here to get his feeling on this.

So, what do you think about all this? Would you be willing to update to the more consistent way 😁?
In that case there is a little catch on the move of content type: there seems to be no translation on the error title somehow

![image](https://user-images.githubusercontent.com/1769475/188761402-1638b765-78ba-446d-bf05-b4ecd450d753.png)

Thanks in advance for the extra help!

Cheers!

CyberReiter commented:
Hi @mikecp 

I also found it strange that the behaviour was different for the different types but I thought I would leave it as it is for now.
However since I am also a fan of consistency I will take another look at it 😁

For the content part I would leave that out for this PR as I don't see a way of changing this without making a breaking change.

CyberReiter commented:
Changed the behaviour to now show an error across the board when moving an item to its parent.

For the missing translations on the content types the problem seems to be that the promise returned by the `localizationservice` is not able to finish in time and thus will always be an empty object. For that reason I went with the same approach as already implemented on the datatypes and just added a static message - since the notification message will still be localized this should be sufficient.

mikecp commented on src/Umbraco.Web.UI.Client/src/common/resources/contenttype.resource.js:
```diff
@@ -432,15 +432,13 @@ function contentTypeResource($q, $http, umbRequestHelper, umbDataFormatter, loca
                         id: args.id
                     }, { responseType: 'text' }),
-                promise);
+              'Failed to move content type');
```
I saw your comment about the promise not being returned on time, but what about keeping the promise and having something like 

`promise ?? 'Failed to move content type'`

This way, if the promise ever gets returned on time, we get the foreseen label and otherwise we have a fallback.

What do you think? (same question in mediatype.resource of course😁 )

CyberReiter replied:
problem is that the promise is not null but an Promise object so this won't work 🤔

There also does not seem to be an easy way of waiting for the promise to finish as there is no way to check for the state of it.

Only possibility that comes to my mind would be to have a bool which would be set to true if the promise finished in time and then set the message based on in it like so `isFinished ? promise : 'Failed to move content type'`
However by doing this we would introduce a 3rd way of handling this message as we already have the way it is done on the data types (static message) and another way for media (static message + taking some of the translation from the api response)
https://github.com/umbraco/Umbraco-CMS/blob/a0d673262ceafc636e4fc5b856b7933733e1fc23/src/Umbraco.Web.UI.Client/src/common/resources/media.resource.js#L130

mikecp replied:
Thanks for your comments @CyberReiter !

In that case, I would suggest to keep things as they are. It's already a great improvement, so it should not hold us from merging all this 😁

Thanks again for this update 👍

Cheers!
